### PR TITLE
Implement the HOMEBREW_INSTALL_CLEANUP env to trigger cleanup on reinstall/install/upgrade

### PR DESCRIPTION
**Pull request by Wojonatior**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

This PR is thethe implementation of the `HOMEBREW_INSTALL_CLEANUP` ENV and associated features as discussed in #4760.

### CODE REVIEW

File: Library/Homebrew/test/cmd/upgrade_spec.rb [hunk 1]
```diff
@@ -17,4 +17,17 @@
     expect(HOMEBREW_CELLAR/"testball/0.1").to be_a_directory
     expect(HOMEBREW_CELLAR/"testball/0.0.1").not_to exist
   end
+
+  it "upgrades a Formula and cleans up old versions when `HOMEBREW_INSTALL_CLEANUP` is set" do
+    setup_test_formula "testball"
+    # allow(ENV).to receive(:[]).and_call_original
+    # allow(ENV).to receive(:[]).with("HOMEBREW_INSTALL_CLEANUP").and_return("1")
+    ENV["HOMEBREW_INSTALL_CLEANUP"] = "1"
```
Comment (Wojonatior) [inline Library/Homebrew/test/cmd/upgrade_spec.rb]: I've tried both manually setting into `ENV` and mocking it out as shown in the 2 commented lines. I wasn't able to get either working even though I could confirm the functionality manually.

File: Library/Homebrew/cmd/upgrade.rb [hunk 2]
```diff
@@ -3,7 +3,7 @@
 #:
 #:    Options for the `install` command are also valid here.
 #:
-#:    If `--cleanup` is specified or `HOMEBREW_UPGRADE_CLEANUP` is set then remove
+#:    If `--cleanup` is specified, `HOMEBREW_INSTALL_CLEANUP` or `HOMEBREW_UPGRADE_CLEANUP` is set then remove
```
Comment (Wojonatior) [inline Library/Homebrew/cmd/upgrade.rb]: As a clarification around the original feature, is the intent to keep both ENVs or to enhance what the old one did as well a rename it?

File: Library/Homebrew/test/cmd/upgrade_spec.rb (hunk 1 above)
Comment (MikeMcQuaid) [reply Library/Homebrew/test/cmd/upgrade_spec.rb]: @Wojonatior Ideally we wouldn't add another integration test for this (they are really slow). Either add a unit test or omit one entirely.

File: Library/Homebrew/cmd/upgrade.rb (hunk 2 above)
Comment (MikeMcQuaid) [reply Library/Homebrew/cmd/upgrade.rb]: @Wojonatior Good question 👍 Let's remove the documentation for `HOMEBREW_UPGRADE_CLEANUP` in favour of `HOMEBREW_INSTALL_CLEANUP` but keep the `HOMEBREW_UPGRADE_CLEANUP` variable functional for people who are still using it.

**Comment by Wojonatior** [comment]
I've added the formula cleanup to the `install` and `reinstall` commands. I wasn't sure if it was supposed to be the formula cleanup or the full/regular cleanup. The issue seems to discuss the formula cleanup, but I'm not sure what would need to be cleaned up from the formula on an `install`/`reinstall`

File: Library/Homebrew/cmd/install.rb [hunk 3]
```diff
@@ -73,6 +73,9 @@
 #:
 #:    If `--git` (or `-g`) is passed, Homebrew will create a Git repository, useful for
 #:    creating patches to the software.
+#:
+#:    If `HOMEBREW_INSTALL_CLEANUP` is set then remove previously installed versions
+#:    of upgraded <formulae>.
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cmd/install.rb]: This also will clean up the HOMEBREW_CACHE for that formula

**Comment by MikeMcQuaid** [comment]
@Wojonatior After fixing the docs please run `brew man` and commit the results.

**Comment by MikeMcQuaid** [comment]
> I wasn't sure if it was supposed to be the formula cleanup or the full/regular cleanup. 

Formula cleanup 👍. Could have regular cleanup occurring every e.g. 100 installs or something, though? Thoughts?

> The issue seems to discuss the formula cleanup, but I'm not sure what would need to be cleaned up from the formula on an `install`/`reinstall`

Nothing specifically but there may be leftover cache and cellar entries that it's worth cleaning up at this stage.

**Comment by Wojonatior** [comment]
@MikeMcQuaid I updated the docs and the manpage. I like the idea of the occasional full clean. Would it make sense to count that in the formula clean or specifically for the install? Is there a pattern established somewhere for how I would persist the number of installs?

**Comment by MikeMcQuaid** [comment]
> Would it make sense to count that in the formula clean or specifically for the install? Is there a pattern established somewhere for how I would persist the number of installs?

I was thinking that rather than being by installs it could be by days. `brew cleanup` could write (or `touch` if it already exists) a file like `HOMEBREW_CACHE/.cleaned` which the `mtime` is checked if `HOMEBREW_INSTALL_CLEANUP` is set and a full, non-formula `cleanup` could be run at that time if it hasn't been run for e.g. 30 days (that's how often `git gc` is run). If we want to be safer still that `brew cleanup` could also only remove files/directories that would be removed anyway _and_ are 30 days old (but that may be overkill).

**Comment by MikeMcQuaid** [comment]
Thanks so much for your first contribution! Without people like you submitting PRs we couldn't run this project. You rock, @Wojonatior! This is useful as-is so figured it was worth merging. Would still love to see the stuff we talked about in a follow-up PR!

**Comment by Wojonatior** [comment]
@MikeMcQuaid Thanks! Sorry about the lack of communication on my part. I definitely want to make the follow-up PR for this, you should hopefully see something around it in the next week or so.

**Comment by MikeMcQuaid** [comment]
@Wojonatior No worries, looking forward to it!

**Pull request merged**